docs/adr/0007-preventing-version-mismatches.md

```diff
@@ -0,0 +1,233 @@
+
+However, this issue seems less important since a) the errors are easier to diagnose and fix and b) we have historically
+only introduced breaking changes like this for "experimental" features like custom/wrapped types.
+
+This ADR will ignore this issue and only focus on compatibility between the generated scaffolding and bindings code.
+
+## Decision Drivers
+
+- UniFFI should have a friendly developer experience.
+  - If possible, developers shouldn't have to worry or know about this issue.
+  - If that's not possible, then it should be easy for them to identify and fix the issue.
+- We need to support external bindings generators and allow those projects to
+  release new versions without coordinating with the UniFFI team.
+
+## Considered Options
+
+### **[Requirement] More flexible versioning**
+
+As mentioned in the context, once we have external bindings generators we can't rely on a simply compare it's version
+string to `uniffi`.  A more flexible versioning policy is a requirement for all other options discussed here.
+
+I think we can use semver for this:
+  - The major version is incremented whenever there's a change that would breaks compatibility between the bindings and
+    scaffolding generation code. Changing the layout of `RustBuffer` or `RustCallStatus` would be a breaking change. Our
+    custom/external types work would not.
+  - For code generators: the minor version is incremented whenever there's a breaking change between the generated code
+    and the consumer code.  Some examples would be: changing Kotlin classes from open to closed and changing the
+    argument names on Swift.
+  - For `uniffi` itself and other crates, I'm not sure what a minor version change should mean.  Maybe we just don't
+    bump these?
+  - The patch version is incremented for non-breaking, bugfix, releases.
+
+This means that `uniffi-bindgen` and all external bindgen tools are compatible with `uniffi` if their major versions
+match.  We would update `ComponentInterface` so it only uses the major version number when calculating it's hash.
+
+Pros:
+  - Seems like a hard requirement if we want external bindings tools
+  - Would significantly reduce the chance of running into a version mismatch.
+  - Makes it easier to detect version mismatches when parsing `Cargo.toml`. Right now we need to determine the exact
+    version that was resolved and stored in `Cargo.lock`.  If we only needed to compare the major versions, then we
+    could use the specification from `Cargo.toml` directly.
+
+Cons
+  - We would have to be more careful with breaking changes. If we realized a change would break the generated code
+    compatibility, then that change might need to wait before releasing it. If we didn't realize a change would break
+    the generated code compatilibity and released the change anyway, then we've introduced undefined behavior in all
+    consumer code.
```

Comment:
As part of implementing this change, we would remove the version number from the calculation of the `ComponentInterface` fingerprint, instead depending on orchestration of the build process to detect and prevent version mismatches?

Reply:
Assuming the above, where the major version of uniffi and the binding must match, the intent would be to only use that major version for this.